File: learning/game_engine/comp4300/Vector2D.cc
#include "Vector2D.h"
#include "cmath"

Vector2D Vector2D::operator + (const Vector2D& rhs) const
{
    return Vector2D{ x + rhs.x, y + rhs.y };
};

Vector2D Vector2D::operator - (const Vector2D& rhs) const
{
    return Vector2D{ x - rhs.x, y - rhs.y };
};

Vector2D Vector2D::operator * (const Vector2D& rhs) const {
    return Vector2D {x * rhs.x, y * rhs.y };
}

Vector2D Vector2D::operator * (const float value) const {
    return Vector2D {x * value, y * value };
}

Vector2D Vector2D::operator / (const Vector2D& rhs) const {
    return Vector2D {x / rhs.x, y / rhs.y };
}

void Vector2D::operator += (const Vector2D& rhs)
{
	x += rhs.x;
	y += rhs.y;
};

void Vector2D::operator -= (const Vector2D& rhs)
{
	x -= rhs.x;
	y -= rhs.y;
};

void Vector2D::operator/= (const Vector2D& rhs) {
    x /= rhs.x;
    y /= rhs.y;    
}


void Vector2D::operator *= (const Vector2D& rhs)
{
	x *= rhs.x;
	y *= rhs.y;
};

Vector2D& Vector2D::Add(float value) {
    x += value;
    y += value;

    return *this;
};

Vector2D& Vector2D::Scale(float scale)
{
	x *= scale;
	y *= scale;

	return *this;
};

Vector2D& Vector2D::Rotate(float angleInDegrees) {
	const float angleInRadians = angleInDegrees * 3.14f / 180.0f;
	const float cosTheta = cos(angleInRadians);
	const float sinTheta = sin(angleInRadians);

	x = x * cosTheta - y * sinTheta;
	y = x * sinTheta + y * cosTheta;

	return *this;
};


Vector2D& Vector2D::Normalize() {
    float len = Length();
    if (len > 0) { // Avoid division by zero
        x /= len;
        y /= len;
    }
    return *this;
}

float Vector2D::Dot(const Vector2D& rhs) const {
    return x * rhs.x + y * rhs.y; 
}

float Vector2D::Length() const
{
	return sqrtf(x * x + y * y);
};

float Vector2D::Distance(const Vector2D& vec2D) const {
    float dx = x - vec2D.x;
    float dy = y - vec2D.y;
    return sqrtf(dx * dx + dy * dy);
}

File: learning/EntityComponentSystem/Vector2D.rb
class Vector2D < Struct.new(:x,:y)

  def initialize(x = 0.0, y = 0.0)
    self.x = x
    self.y = y
  end

  def +(rhs)
    Vector2D.new(x + rhs.x, y + rhs.y)
  end

  def -(rhs)
    Vector2D.new(x - rhs.x, y - rhs.y)
  end

  def *(rhs)
    Vector2D.new(x * rhs.x, y * rhs.y)
  end

  def /(rhs)
    Vector2D.new(x / rhs.x, y / rhs.y)
  end

  def add(value)
    self.x += value
    self.y += value
    self # Return self for chaining
  end

  def scale(scale)
    self.x *= scale
    self.y *= scale
    self
  end

  def rotate(angle_in_degrees)
    angle_in_radians = angle_in_degrees * Math::PI / 180.0
    cos_theta = Math.cos(angle_in_radians)
    sin_theta = Math.sin(angle_in_radians)

    new_x = x * cos_theta - y * sin_theta
    new_y = x * sin_theta + y * cos_theta
    self.x = new_x
    self.y = new_y
    self
  end

  def normalize
    len = length
    if len > 0
      self.x /= len
      self.y /= len
    end
    self
  end

  def dot(rhs)
    x * rhs.x + y * rhs.y
  end

  def length
    Math.sqrt(x * x + y * y)
  end

  def distance(vec2D)
    dx = x - vec2D.x
    dy = y - vec2D.y
    Math.sqrt(dx * dx + dy * dy)
  end
end
